ADR/0033-product-version-controller.md

```diff
@@ -0,0 +1,252 @@
+
+**Note:** The choice was made to use the terms "Project" and "Project
+Development Stream" as opposed to "Product" and "Version" because those can be
+applied more broadly also to upstream community projects as opposed to
+marketable products. Those terms also more accurately describe the environment
+in which software development happens. [See here][pvsp] for a more detailed
+discussion of these terms.
+
+[oref]: https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/
+[pvsp]: https://opensource.com/article/19/11/product-vs-project
+
+### Streamlining version creation from the API/CLI.
+
+We will define the **ProjectDevStreamTemplate** custom resource. The *spec* for
+such a resource will include:
+
+* A **project** element referring to an existing *Project* resource.
+* A **variables** element that includes a list of variable names and optional
+  default values. The default values of variables my be defined in terms of
+  other, previously defined, variables using the Go template syntax.
+* A **resources** element that includes a list of resource definitions so that
+  each definition may reference variables by using the Go template syntax.
+
+Here is an example for what a *ProjectDevStreamTemplate* resource may look like:
+
+```
+apiVersion: pvc.konflux.dev/v1alpha1
+kind: ProjectDevStreamTemplate
+metadata:
+  name: template1
+  ...
+spec:
+  project: project1
+  variables:
+  - name: foo
+  - name: bar
+    defaultValue: "baz{foo}"
+  resources:
+  - apiVersion: appstudio.redhat.com/v1alpha1
+    kind: Application
+    metadata:
+      name: "{foo}-app1"
+    spec:
+      displayName: "{foo}-app1"
+  - apiVersion: appstudio.redhat.com/v1alpha1
+    kind: Component
+    metadata:
+      name: "{foo}-comp1"
+    spec:
+      application: "{foo}-app1"
+      componentName: "{foo}-comp1"
+      source:
+        git:
+          context: ./
+          dockerfileUrl: Dockerfile
+          revision: {bar}
+          uri: ...
+```
+
+The *spec* section for the *ProjectDevStream* may include a **template** section
+that includes a **name** field for referring to a *ProjectDevStreamTemplate*
+resource and a **values** section to provide values for the template variables.
+When this happens, the resources described by the template get created.
+
+Here is an example for what a *ProjectDevStream* resource with a template may
+look like:
+
+```
+apiVersion: pcv.konflux.dev/v1alpha1
+kind: ProjectDevStream
+metadata:
+  ...
+spec:
+  template:
+    name: template1
+    values:
+      - name: foo
+        value: some-value-here
+```
+
+The *values* section must provide values for all template variables without
+default values defined, it may also include values for variable that do have
+default values defined.
+
+The example above will cause the creation of the "some-value-here-app1"
+*Application* resource and the "some-value-here-comp1" *Component* resource.
```

Comment:
What happens if someone tries to edit the components, or say delete one component from a new set belonging to a ProjectDevStream that was created from a template? Does the projectdevstream controller unset what the user did, does it recreate the deleted component?

Or, is this a one-pass thing? i.e., once a project dev stream with a template in its spec has been processed, something updates its status, a label, or an annotation to stop further reconciliation of those resources?

Reply:
I think that implementation-wise it would be simpler to keep the templates and the generated objects in sync via the reconciliation loop. That may also end up being useful for some advanced users (Think: bulk-change multiple versions).

In practice when using the UI, I think each template will only get applied once and discarded.